src/VASSAL/build/module/map/PieceMover.java

```diff
@@ -244,25 +244,16 @@ public Object visitDeck(Deck d) {
       @Override
       public Object visitStack(Stack s) {
         DragBuffer.getBuffer().clear();
-        // RFE 1629255 - Only add selected pieces within the stack to the DragBuffer
-        // Add whole stack if all pieces are selected - better drag cursor
-        int selectedCount = 0;
-        for (int i = 0; i < s.getPieceCount(); i++) {
-          if (Boolean.TRUE.equals(s.getPieceAt(i)
-                                   .getProperty(Properties.SELECTED))) {
-            selectedCount++;
-          }
-        }
+        // RFE 1629255 - Only add selected pieces within the stack to the DragBuffer,
+        // except when global pref "MOVING_STACKS_PICKUP_UNITS" is set
+        final boolean selectAllUnitsInStackRegardlessOfSelection =
+            (Boolean) GameModule.getGameModule().getPrefs().getValue(Map.MOVING_STACKS_PICKUP_UNITS);
 
-        if ((Boolean) GameModule.getGameModule().getPrefs().getValue(Map.MOVING_STACKS_PICKUP_UNITS) || s.getPieceCount() == 1 || s.getPieceCount() == selectedCount) {
-          DragBuffer.getBuffer().add(s);
-        }
-        else {
-          for (int i = 0; i < s.getPieceCount(); i++) {
-            final GamePiece p = s.getPieceAt(i);
-            if (Boolean.TRUE.equals(p.getProperty(Properties.SELECTED))) {
-              DragBuffer.getBuffer().add(p);
-            }
+        for (int i = 0; i < s.getPieceCount(); i++) {
+          final GamePiece p = s.getPieceAt(i);
+          final boolean pieceSelected = Boolean.TRUE.equals(p.getProperty(Properties.SELECTED));
+          if (selectAllUnitsInStackRegardlessOfSelection || pieceSelected) {
```

Comment:
Calling `p.getProperty(Properties.SELECTED)` in the case where we don't need it means we don't get the advantage of short-circuit evaluation. Better would be:

```
if (selectAllUnitsInStackRegardlessOfSelection ||
    Boolean.TRUE.equals(p.getProperty(Properties.SELECTED)) {
```

Reply:
Aside: It would be great to have `Stack` implement `Iterable<GamePiece>` so all the loops which are by index but don't use the index for anything other than calling `getPieceAt()` could just have the `GamePiece` directly.

Reply:
The local boolean variable is there to improve readability, `if (selectAllRegardless || pieceSelected)` reads much better than `if (selectAllRegardless || Boolean.TRUE.methodParenthesisVariableDotMethodParenthesisClassDotConstantParenthesis)`. If it is really necessary to optimize performance at this particular spot without having measured any performance issues here, I can make the requested change, but I ask to reconsider. Code should be written to be read by human developers first and by machines second.

The idea of having the Stack hand out its contents in a standard way already crossed my mind, and I feel the optimal way for this is having the Stack "HAVE A" List of GamePieces (composition) instead of "BE A" List of GamePieces (inheritance). A simple `List<GamePiece> Stack.getContents()` or `getGamePieces()` would solve this in a cleaner way without binding the Stack itself to a new type, and `for (GamePiece piece : aStack)` would look a bit weird while with a `for (GamePiece piece : aStack.getContents())` it would be easier to understand what is going on.

Reply:
On reflection, I think the nicest way to do this would be if `Stack` supported `forEach()`:

```
final DragBuffer d = DragBuffer.getBuffer();
if (selectAllUnitsInStackRegardlessOfSelection) {
  s.forEach((p) -> d.add(p));
}
else {
  s.forEach((p) -> {
    if (Boolean.TRUE.equals(p.getProperty(Properties.SELECTED))) {
      d.add(p);
    }
  );
}
```

Reply:
Even more concise:
```
final DragBuffer d = DragBuffer.getBuffer();
s.forEach(
  selectAllUnitsInStackRegardlessOfSelection ?
  (p) -> d.add(p) :
  (p) -> {
    if (Boolean.TRUE.equals(p.getProperty(Properties.SELECTED))) {
      d.add(p);
    }
  }
);
```

Reply:
For the iterations over the contents of a Stack, I have a proposal for this in PR #23. Why break a leg and have the Stack itself be an Iterable or a foreach-able thing, if the Stack can spit out a List which already has all this?